crates/infra/cli/src/commands/mod.rs

```diff
@@ -67,3 +67,25 @@ impl CLI {
         }
     }
 }
+
+#[test]
+fn generate_zsh_completions() -> Result<()> {
+    use clap::CommandFactory;
+    use clap_complete::shells::Zsh;
+    use infra_utils::cargo::CargoWorkspace;
+    use infra_utils::codegen::Codegen;
+
+    use super::*;
+
+    let mut buffer = vec![];
+    let mut command = CLI::command();
+
+    clap_complete::generate(Zsh, &mut command, "infra", &mut buffer);
+
+    let crate_dir = CargoWorkspace::locate_source_crate("infra_cli")?;
+
+    Codegen::write_only()?.write_file(
+        crate_dir.join("generated/infra.zsh-completions"),
+        String::from_utf8(buffer)?,
+    )
```

Comment:
Why is this a unit test? This looks like a hack and will only work when executing `infra test` if I understand correctly; this should be running as part of the build script or ideally be exposed as a command-line command and executed as part the setup.

Reply:
So, for most of codegen, I originally wrote it to execute via Cargo build tasks, because it affects Cargo rebuild incrementability, as the source code is imported into other crates.
But doing so is not cheap, as any change causes a big part of the workspace graph to be rebuilt, and slowes down local builds and in-editor intellisense/REPL.

But in this case, the output is not consumed by anything else, and it doesn't need to interact with the Cargo build graph, I chose to author it as a unit test. This is similar to CST snapshot tests.
Doing so will not re-evaluate/rebuild it with every key-stroke/file save. Additionally, it rarely changes that we need to consistently see its delta as we edit code.

It still use the `Codegen` API though to make sure the outputs are produced correctly, have the right headers, and are validated (not overwritten) in CI if they are out of date.

Reply:
I understand the reasoning and that it's convenient since it Just Works^TM in this case, but generating observable artifacts as part of the unit test suite is **very** surprising and we're trading one hack for another, arguably worse. Hacks aside, now `cargo check/build` will *not* emit that artifact but surprisingly `cargo test` will.

IMO, it would be cleaner to expose this as a command, i.e. `infra completions zsh > output` and to explicitly use that as part of the devcontainer/`infra` setup, whenever needed. In short:
- it separates what's strictly needed for a correct build (what's usually done as part of the `Codegen`) from optional, auxiliary artifacts
- it's not always beneficial, i.e. Bash users have no need for that file
- as such, it should be up to the user or the "setup" routine to do that on a per-user/environment basis

I'd also argue that there's not much benefit to observe/review changes to the completion files since at that point we'd be testing/reviewing output of the `clap` library ourselves, but that's more of a side-note.

However, if we feel strongly about still always generating this as an artifact of the build, I'd prefer if we used something that expresses the intent and expectations better, specifically [`cargo-post`](https://github.com/phil-opp/cargo-post) for the `infra` crate or try adopting other tools such as [`just`](https://github.com/casey/just) or [`cargo-xtask`](https://github.com/matklad/cargo-xtask).

Reply:
There's a substantial benefit already to getting rid of the duplicate `/npm/` source dir via this PR, so if you feel the above (infra/completion bits) would take more time or effort right now, I'd strongly prefer if we keep the existing behaviour for now and only include the `/npm/` deduplication bits. We can follow-up or brainstorm how we can or should approach handling the completions artifacts better.

What do you think?

Reply:
> as part of the unit test suite is very surprising and we're trading one hack for another, arguably worse.
> it would be cleaner to expose this as a command
> since at that point we'd be testing/reviewing output of the clap library ourselves

Fully agree. But the issue with this approach is that we need to check in that file, so that it is ready when we first load the dev container. Otherwise, the `zsh` terminal won't have it. So the dev experience would be to load the dev container, wait for `infra setup`, then delete/reload `zsh`.

I think it is a bit unconventional, but given that it rarely changes, it seems to me that the test will just be "checking that the file is up to date" in 99.99% of the cases.

However, I'm open for suggestions on how to improve this, especially if we can move it to an `infra` command. Not sure if it is worth blocking/changing this PR for this though. I mainly moved it here to get rid of the path hacks, and to remove the extra binary built/executed needlessly for each time we invoke any other `infra` command.

U

Reply:
I see your point about possibly waiting for `infra setup` to finish. However, like I said, running this as part of a unit test is too big of a hack - we have already complex enough infra/build setup here that adding more obscure indirection such as this is, I feel like, just one step too much. 

There seems to be a `clap_complete` pattern to compile-time generation of the shell completion files, so using a build script to do that is still, I'd say, the most straightforward/readable way to go about it.

I've opened https://github.com/NomicFoundation/slang/pull/739; I cleaned up the build script there, so hopefully it's acceptable in that state. I'll post there another idea for resolving this at run-time, which should get rid of the build script entirely and still provide a good DX. Let me know what you think!

Reply:
Alright. I will revert that bit from this PR for now, so that we can land it, and we can follow up on the completions issue on your PR 👍

Reply:
Thanks for understanding :heart:  I'm happy to follow up on this issue in #739.